### PR TITLE
Add --mypy-xfail

### LINKED ISSUE #177: Option to treat mypy failures as warnings
In jaraco/skeleton#143, I'm attempting to adopt more strict mypy settings for my projects. Because these checks are part of the test suite and CI checks, any failed check breaks the integration.

I'd like a way to treat mypy errors as non-blocking, to get a report of failures but not fail the check, similar to how warnings work in pytest. It would be nice if pytest-mypy could expose an option to do something similar.

I could see a number of ways to approach this need:

- Have errors emitted as actual warnings.
- Disable failures, but provide a separate report (similar to pytest-cov).
- Treat the failures as xfail.

Maybe there are other options too.

### PR DESCRIPTION
**Pull request by dmtucker**
Resolve #177 

> I could see a number of ways to approach this need:
>
> - Have errors emitted as actual warnings.
> - Disable failures, but provide a separate report (similar to pytest-cov).
> - Treat the failures as xfail.

I spent some time with each of these and ruled out emitting warnings for the following reasons.
- Warnings are not pretty. Here's what my testing looked like:
  ``` console
  demo/bad/__init__.py::mypy1.11.1
    /home/dtux/Projects/pytest-mypy/src/pytest_mypy/__init__.py:237: MypyWarning: 1: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
      warnings.warn(MypyWarning(file_error_formatter(self, results, errors)))
  ```
  - `/home/dtux/Projects/pytest-mypy/src/pytest_mypy/__init__.py:237` ...don't care
  - `warnings.warn(MypyWarning(file_error_formatter(self, results, errors)))` ...don't care
- Moreover, IIUC, warnings not really appropriate for end-users because they are aimed at developers (hence the info we don't care about above).
  - On that note, we should probably do something about this: https://github.com/realpython/pytest-mypy/blob/23131baea4afbbae158bacd77955cc967903989c/src/pytest_mypy/__init__.py#L238

This implements the latter 2 bullets via a new CLI option: `--mypy-xfail`.
- If a mypy check (including the status check) passes, Pytest reports a pass:
  ``` console
  $ venv/bin/pytest --mypy-xfail demo/good.py 
  ====================== test session starts =======================
  platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
  Using --randomly-seed=3285518892
  rootdir: /home/dtux/Projects/pytest-mypy
  configfile: tox.ini
  plugins: cov-4.1.0, mypy-0.10.4.dev48+gb3a6274.d20240813, randomly-3.15.0, xdist-1.34.0, forked-1.6.0
  collected 2 items                                                

  demo/good.py ..                                            [100%]
  ============================== mypy ==============================
  Success: no issues found in 1 source file

  ======================= 2 passed in 0.09s ========================
  ```
- If a check fails, Pytest reports an xfail (with reasonable reasons, shown with `-ra`):
  ``` console
  $ venv/bin/pytest --mypy-xfail -ra demo/bad.py
  ====================== test session starts =======================
  platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
  Using --randomly-seed=2239210653
  rootdir: /home/dtux/Projects/pytest-mypy
  configfile: tox.ini
  plugins: cov-4.1.0, mypy-0.10.4.dev48+gb3a6274.d20240813, randomly-3.15.0, xdist-1.34.0, forked-1.6.0
  collected 2 items                                                
  
  demo/bad.py xx                                             [100%]
  ============================== mypy ==============================
  demo/bad.py:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
  Found 1 error in 1 file (checked 1 source file)

  ==================== short test summary info =====================
  XFAIL demo/bad.py::mypy1.11.1 - The mypy errors in this file were xfailed by --mypy-xfail.
  XFAIL demo/bad.py::mypy1.11.1-status - mypy's non-zero exit status (1) was xfailed by --mypy-xfail.
  ======================= 2 xfailed in 0.28s =======================
  ```

### CODE REVIEW

File: src/pytest_mypy/__init__.py [hunk 1]
```diff
@@ -253,6 +267,14 @@ def runtest(self) -> None:
...
         results = MypyResults.from_session(self.session)
         if results.status:
+            if self.session.config.option.mypy_xfail:
+                self.add_marker(
+                    pytest.mark.xfail(
+                        raises=MypyError,
+                        reason=f"mypy's non-zero exit status ({results.status})"
+                        " was xfailed by --mypy-xfail.",
+                    )
+                )
```
Comment (jaraco) [inline src/pytest_mypy/__init__.py:277]: Nit: This code feels a little redundant with the earlier code. Maybe extract a method?

File: src/pytest_mypy/__init__.py [hunk 2]
```diff
@@ -234,6 +240,14 @@ def runtest(self) -> None:
                 error.partition(":")[2].partition(":")[0].strip() == "note"
                 for error in errors
             ):
+                if self.session.config.option.mypy_xfail:
+                    self.add_marker(
+                        pytest.mark.xfail(
+                            raises=MypyError,
+                            reason="The mypy errors in this file"
+                            " were xfailed by --mypy-xfail.",
```
Comment (jaraco) [inline src/pytest_mypy/__init__.py:248]: Nit: The wording feels a little verbose, especially considering pytest also emits "XFAIL" in the output. Maybe rephrase for conciseness?

```suggestion
                            reason="Expected mypy errors in this file.",
```
or
```suggestion
                            reason="Expected mypy errors in this file (--mypy-xfail).",
```

(similar for status)

**Review by jaraco — APPROVED** [review]
Thanks for the investigation and explanation. This approach makes sense and is not too complicated.

I do worry that with this feature, I or others will want finer-grained control (only specific classes of failure or specific files). Still, for the application I was seeking, this would have worked great. I could have added the xfail directive to all of the projects, rolled out the stricter changes, and then project-by-project addressed the expected failures until they subsided to then remove the directive.

File: src/pytest_mypy/__init__.py (hunk 1 above)
Comment (dmtucker) [reply src/pytest_mypy/__init__.py:277]: Think I'm gonna punt on this for now... I've got a branch that explores using different errors for these checks, and that would change what gets passed to `raises` (which I think would make a method less useful).
